README.md
# OVHcloud Web Statistics

# Introduction
OWstats is a webanalytics service which reports website usage (traffic, users localization...) to OVHcloud webhosting customers thanks to server logs analysis. There are 3 building blocks to this service :
- owstats-backend : OVHcloud webhosting Apache servers logs are parsed, transformed and the result is stored in databases
- owstats-api : databases built with owstats-backend are exposed through owstats-api
- owstats-ui : frontend part of the application is displaying data to final users

This service is already installed and available for all OVHcloud hostings (see https://docs.ovh.com/ie/en/hosting/shared_view_my_websites_logs_and_statistics/). 
The aim of this open source repository is to provide more flexibilty to users who want to analyse deeper their server logs.

Currently, only owstats-ui code is open source. It is planned to release backend and api code in the near future so owstats can be fully installed on other servers (and not limited to OVHcloud webhosting customers). 

# A. Generate OVHcloud API credentials
In order to use owstats-ui, you will need to have OVHcloud API credentials. You can generate them by following this documentation :
 - https://docs.ovh.com/ie/en/api/first-steps-with-ovh-api/#advanced-usage-pair-ovhcloud-apis-with-an-application_2
 - in createToken page, set validity as "Unlimited" and rights as "GET" "/hosting/web*" (see picture below)
 - Script name and description can have any value you wish
 - note down Application Key, Application Secret and Consumer Key generated.

![credentials](./docs/doc-app-keys.png)

# B. Owstats-ui local development

## Pre-requesites
- git : https://ovh.github.io/manager/how-to/#install-git
- nvm (or node v12, v14 or v16) : https://ovh.github.io/manager/how-to/#install-node-js
- an OVHcloud Web Hosting plan (https://www.ovh.ie/web-hosting/)
- OVHcloud API credentials generated (see "Generate OVHcloud API credentials" section)

## Local developpement configuration

In your local machine, install dependencies and launch development server by doing the following commands in a terminal :

```
nvm install 16
nvm use 16
npm install --global yarn
git clone https://github.com/ovh/owstats-ui.git
cd owstats-ui
yarn install
yarn local-config
yarn serve
```

Then, you can access your local logs website (replace hostingname by your hosting) :
```
http://localhost:8080/hostingname/owstats#/dashboard
```

## Launch cypress (front end testing dashboard)

Be sure to have no process running on port 8090 of your localhost and then run this command in terminal. It will open up a dashboard where you can display owstats-ui with test data :

```
yarn cypress-dashboard
```

See cypress documentation for more information : https://docs.cypress.io/

Troubleshooting : if you encounter the error "No version of Cypress is installed in:" you can install cypress manually with the command "./node_modules/.bin/cypress install" (else, other installation troubleshooting and methods: https://docs.cypress.io/guides/getting-started/installing-cypress#System-requirements)

## Other yarn commands

### Compiles and minifies for production
```
yarn build
```

### Lints and fixes files
```
yarn lint
```

### Tests
```
yarn test
```

# C. Deploy owstats UI

You can use owstats-ui repository to deploy your own webanalytics service. For now, this functionality is limited to display OVHcloud webhosting data. It is planned to release API and backend code in order to install full owstats stack for other servers data (VPS...).

For deployment, you need to understand the key distinction between "hosting" and "domain name" :
- hosting: your OVHcloud webhosting service. Data displayed in your webanalytics service will be data related to this webhosting.
- domain name: the domain name where your webanalytics service will be accessible (url). Data displayed in your webanalytics service will not depend on the domain name chosen, only on the hosting.
- after deployment, your webanalytics service will be accessible at the url "{domain}/{hosting}/owstats/#/"

Another important point to understand is who can access your webanalytics service after deployment. Your webanalytics service will be protected by username and password. These credentials are the ones administrated in "User administration" section of "Statistics and Logs" panel in OVHcloud webhosting manager ( https://docs.ovh.com/fr/hosting/mutualise-consulter-les-statistiques-et-les-logs-de-mon-site/#administration-des-utilisateurs )

## Pre-requesites
- git : https://ovh.github.io/manager/how-to/#install-git
- nvm (or node v12, v14 or v16) : https://ovh.github.io/manager/how-to/#install-node-js
- an OVHcloud Web Hosting plan (https://www.ovh.ie/web-hosting/)
- OVHcloud API credentials generated (see "Generate OVHcloud API credentials" section)

## Deploy: Domain name attached to OVHcloud hosting

The most simple option is to deploy your webanalytics service on a domain name attached to your OVHcloud webhosting. In order to do so, perfom following commands in your local machine terminal :

```
nvm install 16
nvm use 16
npm install --global yarn
git clone https://github.com/ovh/owstats-ui.git
cd owstats-ui
yarn install
yarn deploy
```

Then, select "OVHcloud Webhosting" when asked the question :

```
Do you want your analytics service to be published at an url associated to your OVHcloud hosting or at another url ?
```

Follow the instructions and your webanalytics service will automatically be deployed.

## Deploy: Domain name in other hosting services

You also have the possibility to deploy your webanalytics service on a domain name outside OVHcloud webhosting. Be aware that :
- this choice will affect only the url of your analytics service, data displayed will still be computed from the logs of your OVHcloud webhosting
- it requires more manual actions than deployment on an OVHcloud hosting domain name

```
nvm install 16
nvm use 16
npm install --global yarn
git clone https://github.com/ovh/owstats-ui.git
cd owstats-ui
yarn install
yarn deploy
```

Then, select "Other server" when asked the question :
```
Do you want your analytics service to be published at an url associated to your OVHcloud hosting or at another url ?
```

Folder "deploy" is then created in owstats-ui directory. This folder contains all the files needed to deploy your webanalytics service. You need to upload them to the server which serves your domain name :
- ".env.owstats" : it stores your OVHcloud API credentials : be careful to not expose it publicly
- "login_owstats.php" : it must be accessible at the address "{domain}/login_owstats.php" and its variable "$location" must point to ".env.owstats" file location (by default it is "../.env.owstats")
- "root/{hosting}/owstats" folder: files in this folder must be accessible at the address "{domain}/{hosting}/owstats/{file_path}". 

- For example : your domain name is "test.com" and your OVHcloud webhosting is "hosting.ovh". After running the command "yarn deploy', "countries.geojson" file which is in directory "deploy/root/hosting.ovh/owstats". After manual deployment, "countries.geojson" must be accessbile at the address "test.com/hosting.ovh/owstats/data/countries.json"

# D. Modify owstats-ui
You can modify the code of owstats-ui to display additional data or change the styles :
- it uses vuejs, vue router and vuex
- You can use any raw statistics that are documented in owstats API documentation : https://ovh.github.io/owstats-ui/
- styles (colors, spacing...) can be customized easily using the file "src/assets/sass/_variables.scss"
- after modification, you can dislay your changes locally (section B) or deploy your analytics service (section C)

debian/changelog
owstats-ui (1.2.1) unstable; urgency=high

  * Fix bug: get domain list was failing (times span > 10 years)

 -- Lucas Chaigne <lucas.chaigne@corp.ovh.com>  Fri, 06 Jan 2022 15:18:38 +0200

owstats-ui (1.2.0) unstable; urgency=medium

  * Add CDN data

 -- Lucas Chaigne <lucas.chaigne@corp.ovh.com>  Wed, 29 Aug 2022 15:18:38 +0200

owstats-ui (1.1.0) unstable; urgency=medium

  * Support node v14 and v16
  * Possibility to deploy custom web statistics easily
  * Add unit tests and integration tests
  * Bug fixing : average session time computation corrected
  * Bug fixing : domain info displayed if data only in pages and not in visits
  * Finish code reorganisation (split large functions and components)
  * Update documentation for open source : README, CONTRIBUTORS, CONTRIBUTING, LICENSE

 -- Lucas Chaigne <lucas.chaigne@corp.ovh.com>  Wed, 1 Dec 2021 15:13:38 +0200
owstats-ui (1.0.0-1) unstable; urgency=medium

  * Reorganize folders and files
  * Move API calls in store
  * OVH branding (colors, logo, fonts...)
  * Domain selection bug fixing : dropdown list displays subdomains
  * TopDomains bug fixing : component displayed when no error page

 -- Lucas Chaigne <lucas.chaigne@corp.ovh.com>  Wed, 30 Jun 2021 17:26:38 +0200

owstats-ui (0.6.0) unstable; urgency=medium

  * Update UI with OVH branding

 -- Matias Hastaran <matias.hastaran@corp.ovh.com>  Fri, 22 Jan 2021 11:19:58 +0100

owstats-ui (0.5.0) unstable; urgency=medium

    * Add FAQ
    * Add temporary link for the open beta

 -- Matias Hastaran <matias.hastaran@corp.ovh.com>  Fri, 25 Sep 2020 12:58:42 +0200

owstats-ui (0.4.2) UNSTABLE; urgency=medium

    * Fix Dashborad traffic tooltip
    * Add tooltips

 -- Matias Hastaran <matias.hastaran@corp.ovh.com>  Tue, 07 Apr 2020 10:40:21 +0200

owstats-ui (0.4.1) UNSTABLE; urgency=low

    * New build workflow

 -- Matias Hastaran <matias.hastaran@corp.ovh.com>  Mon, 01 Apr 2020 14:00:00 +0100

owstats-ui (0.3.1) UNRELEASED; urgency=low

    * Backend: Fix error on malformed logs entries

 -- Matias Hastaran <matias.hastaran@corp.ovh.com>  Mon, 03 Mar 2020 11:18:00 +0100

owstats-ui (0.3.0) UNRELEASED; urgency=low

    * Alpha version of the UI.

 -- Matias Hastaran <matias.hastaran@corp.ovh.com>  Mon, 02 Mar 2020 10:13:00 +0100

owstats-ui (0.2.0-1) UNRELEASED; urgency=low

    * Initial UI packaging

 -- Tarik Megzari <tarik.megzari@corp.ovh.com>  Tue, 25 Feb 2020 17:30:00 +0100


src/services/utils.js
import moment from 'moment-timezone'
import Cookies from 'js-cookie'

export default {

  // reverse an object of the form { key: value } (ex : { Chrome: 10, Firefox: 10, Opera: 30 })
  // result is the reverse object (data) and its ordered keys (keys)
  // ex of result : { keys:[30, 10],  data: { 10: [Chrome, Firefox], 30: [Opera] } }
  reversePerKeyObject (perKey, sort = 'DESC') {
    const data = {}

    for (const key in perKey) {
      const v = perKey[key]
      if (v in data) {
        data[v].push(key)
      } else {
        data[v] = [key]
      }
    }

    const keys = Object.keys(data)

    if (sort === 'ASC') {
      keys.sort(function (a, b) { return b - a })
    } else {
      keys.sort(function (a, b) { return a - b })
    }
    keys.reverse()

    return { keys, data }
  },

  // return the type of data sent by Api (example: "visits" for v1/initial_path/visits endpoint )
  apiResponseDataType (response) {
    let responseType = ''

    if (Object.keys(response).length > 0) {
      const firstDate = Object.keys(response)[0]
      const key = Object.keys(response[firstDate])[0]
      responseType = key
    }
    return responseType
  },

  // from a list of columns, return the set of columns which will be the key data will be aggregated on
  // standard columns are id, total, ratio, key columns are the other columns
  // if only one column is returned it will be the key (example: Browser)
  computeKeyColumns (columns) {
    const regularColumns = ['id', 'count', 'ratio']
    const keyColumns = []

    const columnsNames = []
    columns.forEach(c => columnsNames.push(c.key))

    columnsNames.forEach(k => {
      if (!regularColumns.includes(k)) {
        keyColumns.push(k)
      }
    })

    return keyColumns
  },

  // replace value for elements of raw data (ex : NT --> Microsoft Windows)
  // columnName : String, name of the column for which values will be renamed
  // record : Object, record for which value will be replaced
  // value : String, new value to be used
  dataValueReplacement (columnName, record) {
    let value = record[columnName]
    let version

    switch (columnName) {
      case 'platform':
        version = record.platform_version
        switch (value) {
          case 'NT':
            value = 'Microsoft Windows'
            break
          case 'X11':
            if (version.includes('Ubuntu')) {
              value = 'Ubuntu'
            } else if (version.includes('FreeBSD')) {
              value = 'FreeBSD'
            } else {
              value = 'Linux'
            }
            break
          case 'Linux':
            if (version.includes('Android')) {
              value = 'Android'
            }
            break
          case 'Macintosh':
            value = 'Mac OS'
            break
        }
        break
      case 'browser':
        if (value === 'MSIE') {
          value = 'Internet Explorer'
        }
        break
      case 'browser_version':
        version = parseFloat(value)
        if (version) {
          value = Math.floor(version).toString()
        }
        break
      case 'country':
        if (value === 'Russian Federation') {
          value = 'Russia'
        } else if (value === 'United States of America') {
          value = 'United States'
        }
        break
    }

    return value
  },

  // given key columns and aggregation key, get the values to be displayed in plain table for key columns
  // aggregationKey: string, key columns values sepatated by " - ", ex : "Microsoft Windows - Chrome - 83"
  // columns: Array, key columns to be displayed in plain table, ex : "platform, browser, version"
  // keyColumnsDiplay : object to display key in separated columns in table
  // ex of keyColumsDisplay: {"platform": "Android", "browser": "Chrome", "browser_version": "34"}
  getRecordKeyColumnsDisplay (aggregationKey, columns) {
    const aggregationKeyArray = aggregationKey.split(' - ')
    const keyColumns = this.computeKeyColumns(columns)
    const keyColumnsDisplay = {}
    for (const i in keyColumns) {
      const column = keyColumns[i]
      let key
      aggregationKeyArray[i] ? key = aggregationKeyArray[i] : key = null
      keyColumnsDisplay[column] = key
    }
    return keyColumnsDisplay
  },

  // given a record and a list of key columns, get the correspoding aggregation key
  // record : Object corresponding to one entry of raw data (ex: {"browser": "Chrome", "browser_version": "83.0.4099.2", "platform": "NT", "platform_version": "10.0", "value": "1"}})
  // columns: list of string: key columns of the plain table (ex: [ "platform", "browser", "browser_version" ])
  // aggregation key : string, aggregation of the values of the key columns for the record, separated by "-", for example : "Microsoft Windows - Chrome - 83"
  getRecordAggregationKey (record, columns) {
    let aggregationKey = ''
    const keyColumns = this.computeKeyColumns(columns)

    // code specific for tables which are more complex than key/total/ratio
    if (keyColumns.length > 1) {
      for (const j in keyColumns) {
        const columnName = keyColumns[j]
        const columnValue = this.dataValueReplacement(columnName, record)

        if (aggregationKey !== '') {
          aggregationKey = aggregationKey.concat(' - ')
        }
        aggregationKey = aggregationKey.concat(columnValue)
      }
      // code specific for simple table (key/total/ratio)
    } else {
      aggregationKey = this.dataValueReplacement(keyColumns[0], record)
    }

    return aggregationKey
  },

  // from raw owstats-api response, key columns and column on which sum is to be done, compute the sum of values by key
  // rawData: Object, api response for which data will be aggregated
  // columns: list of string: key columns of the plain table (ex: [ "browser" ])
  // aggregationOnColumn : string, name of the column on which the sum will be performed (ex: "value")
  // aggregatedDataPerKey: Object, result of the aggregation, for example : { Safari: 10, Chrome : 15 }
  // sumValues : Int, total of the values of the aggregated data (ex : 25)
  // keyColumnsDisplay : if the key is composed of more than one column, object to display key in separated columns in table
  computeAggregationSum (rawData, columns, aggregationOnColumn) {
    let sumValues = 0
    const aggregatedDataPerKey = {}
    const responseDataType = this.apiResponseDataType(rawData)
    const keyColumnsDisplay = {}

    for (const date in rawData) {
      for (const i in rawData[date][responseDataType]) {
        const record = rawData[date][responseDataType][i]
        const onColumnValue = parseInt(record[aggregationOnColumn])
        const aggregationKey = this.getRecordAggregationKey(record, columns)
        keyColumnsDisplay[aggregationKey] = this.getRecordKeyColumnsDisplay(aggregationKey, columns)

        if (aggregationKey in aggregatedDataPerKey) {
          aggregatedDataPerKey[aggregationKey] += onColumnValue
        } else {
          aggregatedDataPerKey[aggregationKey] = onColumnValue
        }
        sumValues += onColumnValue
      }
    }

    return { aggregatedDataPerKey, sumValues, keyColumnsDisplay }
  },

  // from raw owstats-api response, key columns and columns on which average is to be done, compute the avergage of values by key
  // rawData: Object, api response for which data will be aggregated
  // columns: list of string: key columns of the plain table (ex: [ "country" ])
  // aggregationOnColumn : string, name of the column on which the sum will be performed (ex: "value")
  // aggregationByColumn : string, name of the column by which the sum will be divided to get the avergage (ex: "hits")
  // aggregatedDataPerKey: Object, result of the aggregation, for example : { France: 210, Belgium : 234 }
  // sumValues : null, this metric is not needed fot average
  // keyColumnsDisplay : if the key is composed of more than one column, object to display key in separated columns in table
  computeAggregationAverage (rawData, columns, aggregationOnColumn, aggregationByColumn) {
    const aggregatedDataPerKey = {}
    const responseDataType = this.apiResponseDataType(rawData)
    const temporaryAggregatedData = {}
    const keyColumnsDisplay = {}

    for (const date in rawData) {
      for (const i in rawData[date][responseDataType]) {
        const record = rawData[date][responseDataType][i]
        const onColumnValue = parseInt(record[aggregationOnColumn])
        const byColumnValue = parseInt(record[aggregationByColumn])

        const aggregationKey = this.getRecordAggregationKey(record, columns)
        keyColumnsDisplay[aggregationKey] = this.getRecordKeyColumnsDisplay(aggregationKey, columns)

        if (aggregationKey in temporaryAggregatedData) {
          temporaryAggregatedData[aggregationKey].onColumnValue += onColumnValue
          temporaryAggregatedData[aggregationKey].byColumnValue += byColumnValue
        } else {
          temporaryAggregatedData[aggregationKey] = {}
          temporaryAggregatedData[aggregationKey].onColumnValue = onColumnValue
          temporaryAggregatedData[aggregationKey].byColumnValue = byColumnValue
        }
      }
    }

    for (const key in temporaryAggregatedData) {
      const responseTime = (temporaryAggregatedData[key].onColumnValue / temporaryAggregatedData[key].byColumnValue).toFixed(2)
      aggregatedDataPerKey[key] = responseTime
    }

    return { aggregatedDataPerKey, sumValues: null, keyColumnsDisplay }
  },

  // from API response, aggregate data for tables and charts display
  // rawData: Object, api response for which data will be aggregated
  // columns: Array, columns to be displayed in chart/table : aggregation key will be computed with these columns
  // aggregatedDataPerKey: Object, result of the aggregation, for example : { Safari: 10, Chrome : 15 }
  // sumValues : Int, total of the values of the aggregated data (ex : 25)
  // keyColumnsDisplay : if the key is composed of more than one column, object to display key in separated columns in table
  // ex of keyColumsDisplay: { "Android - Chrome - 34 : {"platform": "Android", "browser": "Chrome", "browser_version": "34"},  ...}
  rawDataAggregation (rawData, columns, aggregationType = { type: 'sum', onColumn: 'value' }) {
    if (aggregationType.type === 'sum') {
      const { aggregatedDataPerKey, sumValues, keyColumnsDisplay } = this.computeAggregationSum(rawData, columns, aggregationType.onColumn)
      return { aggregatedDataPerKey, sumValues, keyColumnsDisplay }
    } else {
      const { aggregatedDataPerKey, sumValues, keyColumnsDisplay } = this.computeAggregationAverage(rawData, columns, aggregationType.onColumn, aggregationType.byColumn)
      return { aggregatedDataPerKey, sumValues, keyColumnsDisplay }
    }
  },

  // from aggregated data, format data for chart and table display
  // aggregatedData : Object, result of rawDataAggregation functionÒ
  // columns: Array, columns to be displayed in chart/table : aggregation key will be computed with these columns
  // tableSize: Int, max number of rows to be displayed in the table
  // hasChart: boolean, true if chart data is computed (if component displays a chart)
  // chartSize: Int, max number of keys to be displayed in the pie chart (other keys are aggregated in "other" section)
  // tableData: Array, each element is an object representation of a row to be displayed in the table component
  // labels: Array, each element is a string with the label to be displayed in pie chart component. Computed only if hasChart is true
  // values: Array, each element is a number with the value to be displayed in pie chart component. Computed only if hasChart is true
  computeTableAndChartData (aggregatedData, columns, tableSize, hasChart, chartSize, sort = 'DESC', formatFunction) {
    const { aggregatedDataPerKey, sumValues, keyColumnsDisplay } = aggregatedData
    const { keys, data } = this.reversePerKeyObject(aggregatedDataPerKey, sort)
    const keyColumns = this.computeKeyColumns(columns)

    let others = sumValues
    let id = 0
    let ratio = 0

    const values = []
    const labels = []
    const tableData = []

    for (const i in keys) {
      const key = keys[i]
      const entries = data[key]
      for (const j in entries) {
        const entry = entries[j]
        id += 1
        ratio = (key * 100) / sumValues
        if (hasChart && id <= chartSize) {
          labels.push(entry)
          values.push(parseInt(key))
          others -= key
        }

        const table = {
          id: id,
          count: formatFunction ? formatFunction(key) : key,
          ratio: `${ratio.toFixed(1)}`
        }
        if (keyColumnsDisplay) {
          keyColumns.forEach((e) => {
            table[e] = keyColumnsDisplay[entry][e]
          })
        } else {
          table[keyColumns[0]] = entry
        }

        tableData.push(table)
        if (id >= tableSize) {
          break
        }
      }

      if (id >= tableSize) {
        break
      }
    }

    if (hasChart && others) {
      labels.push('Others')
      values.push(others)
    }

    if (hasChart) {
      return { labels, values, tableData }
    } else {
      return tableData
    }
  },

  // set moment to UTC by default and standardize date formats
  initializeMoment () {
    moment.tz.setDefault('UTC')
    const dayFormat = 'YYYY-MM-DD'
    const weekFormat = 'GGGG-[W]WW' /* Use ISO year GGGG with ISO week WW */
    const monthFormat = 'YYYY-MM'
    const yearFormat = 'YYYY'
    const timeFormat = 'YYYY-MM-DDTHH:mm:ss[Z]'

    return { dayFormat, weekFormat, monthFormat, yearFormat, timeFormat }
  },

  // from start and end date, choose to display in hours, days... intervals
  computeGranularity (start, end) {
    let granularity = 'hours'

    const diff = end.diff(start, 'days')
    if (diff >= 365) {
      granularity = 'months'
    } else if (diff >= 180) {
      granularity = 'weeks'
    } else if (diff >= 30) {
      granularity = 'days'
    }
    return granularity
  },

  computeChartYValue (records, period, type, time, filterParameter) {
    let y = 0
    const r = records[period][type]
    let record

    if (filterParameter) {
      record = r.find(e => (e.time === time && e[filterParameter.key] === filterParameter.value))
    } else {
      record = r.find(e => e.time === time)
    }
    if (record) y = parseInt(record.value)

    return y
  },

  // compute x and y data for spline chart
  // startDate: string, beginning of x axis
  // endDate: String, end of x axis
  // records: Object, raw data to be displayed in the chart
  // filterParameter: optional, Object of format {key:param, value:hits} : filter only records with param equal to hits
  // chartData: Object, data to be displayed in the chart
  // axisData: Array, points of the x axis
  // granularity: string, represents the interval between 2 points of the x axis (hours, days...)
  computeSplineChartData (startDate, endDate, records, filterParameter) {
    moment.tz.setDefault('UTC')
    const { dayFormat, weekFormat, monthFormat, yearFormat, timeFormat } = this.initializeMoment()
    const type = this.apiResponseDataType(records)
    const axisData = []
    const chartData = []
    const stats = {}
    const start = startDate
    const end = endDate
    let period = 'hours'

    const granularity = this.computeGranularity(startDate, moment(end))

    for (let m = moment(start); m.diff(end, 'days') <= 0; m.add(1, period)) {
      const day = m.format(dayFormat)
      const week = m.format(weekFormat)
      const month = m.format(monthFormat)
      const year = m.format(yearFormat)
      const time = m.format(timeFormat)

      let x = time
      let y = 0

      if (granularity === 'days') {
        x = m.format('DD/MM/YYYY')
        x = moment(x, 'DD/MM/YYYY').tz('UTC').format('MM/DD/YYYY')
      } else if (granularity === 'weeks') {
        x = m.format(weekFormat)
        x = moment(x, weekFormat).tz('UTC').format('MM/DD/YYYY')
      } else if (granularity === 'months') {
        x = m.format(monthFormat)
        x = moment(x, monthFormat).tz('UTC').format('MM/DD/YYYY')
      } else {
        x = m.format('DD/MM/YY HH[h]')
        x = moment(x, 'DD/MM/YY HH[h]').tz('UTC').toDate()
      }

      if (day in records) {
        period = 'hours'
        y = this.computeChartYValue(records, day, type, time, filterParameter)
      } else if (week in records) {
        period = 'hours'
        y = this.computeChartYValue(records, week, type, time, filterParameter)
      } else if (month in records) {
        period = 'days'
        y = this.computeChartYValue(records, month, type, time, filterParameter)
      } else if (year in records) {
        period = 'days'
        y = this.computeChartYValue(records, year, type, time, filterParameter)
      }
      stats[x] || (stats[x] = 0)
      stats[x] += y
    }
    Object.keys(stats).forEach(x => {
      axisData.push(x)
      chartData.push(stats[x])
    })

    return { chartData, axisData, granularity }
  },

  // compute average session time over a period of time
  // avgSessionTime : object, result of hour/avgsessiontime api call
  // visits: object, result of hour/visits api call
  // startDate: string, beginning of time period
  // endDate: String, end of time period
  // avgSessionTimeString: string, result of computation in string format (ex: 5mn 30sec)
  computeAvgSessionTime (avgSessionTime, visits, startDate, endDate) {
    const { dayFormat, weekFormat, monthFormat, yearFormat, timeFormat } = this.initializeMoment()

    let period = 'hours'
    const start = startDate
    const end = endDate

    const visitsAndTotalTime = {
      visits: 0,
      totalTime: 0
    }

    for (let m = moment(start); m.diff(end, 'days') <= 0; m.add(1, period)) {
      const day = m.format(dayFormat)
      const week = m.format(weekFormat)
      const month = m.format(monthFormat)
      const year = m.format(yearFormat)
      const time = m.format(timeFormat)

      if (day in avgSessionTime) {
        period = 'hours'
        this.updateVisistsAndTotalTime(avgSessionTime, visits, day, time, visitsAndTotalTime)
      } else if (week in avgSessionTime) {
        period = 'hours'
        this.updateVisistsAndTotalTime(avgSessionTime, visits, week, time, visitsAndTotalTime)
      } else if (month in avgSessionTime) {
        period = 'days'
        this.updateVisistsAndTotalTime(avgSessionTime, visits, month, time, visitsAndTotalTime)
      } else if (year in avgSessionTime) {
        period = 'days'
        this.updateVisistsAndTotalTime(avgSessionTime, visits, year, time, visitsAndTotalTime)
      }
    }

    let avgSessionTimeResult = Math.round(visitsAndTotalTime.totalTime / visitsAndTotalTime.visits)
    if (Number.isNaN(avgSessionTimeResult)) {
      avgSessionTimeResult = 0
    }
    if (avgSessionTimeResult > 60) {
      const durationValue = moment.duration(avgSessionTimeResult, 'seconds')
      const minutes = Math.floor(durationValue.asMinutes())
      const seconds = Math.floor(durationValue.asSeconds() - minutes * 60)
      const avgSessionTimeString = minutes + 'min ' + seconds + 'sec'
      return avgSessionTimeString
    } else {
      const avgSessionTimeString = avgSessionTimeResult + 'sec'
      return avgSessionTimeString
    }
  },

  // function to update visitsAndTotalTime during computeAvgSessionTime
  // avgSessionTime : object, result of hour/avgsessiontime api call
  // visits: object, result of hour/visits api call
  // period: either a day (ex: 2021-01-01), a week (2021-W1), a month (2021-01) or a year (2021)
  // time: time of the for loop in computeAvgSessionTime (ex: 2021-01-01T01:00:00Z)
  // visitsAndTotalTime: object to update with new values ( ex: {visits: 12, totalTime: 150} )
  updateVisistsAndTotalTime (avgSessionTime, visits, period, time, visitsAndTotalTime) {
    let avgSessionTimeValue
    let visitsValue

    const objectAvgSessionTime = avgSessionTime[period].avgsessiontime
    const recordAvgSessionTime = objectAvgSessionTime.find(elem => elem.time === time)
    if (recordAvgSessionTime) avgSessionTimeValue = parseInt(recordAvgSessionTime.value)

    const objectVisits = visits[period].visits
    const recordVisits = objectVisits.find(elem => elem.time === time)
    if (recordVisits) visitsValue = parseInt(recordVisits.value)

    if (visitsValue != null && avgSessionTimeValue != null) {
      visitsAndTotalTime.visits += visitsValue
      visitsAndTotalTime.totalTime += visitsValue * avgSessionTimeValue
    }
  },

  formatBytes (bytes, decimals = 2) {
    if (bytes === 0) return '0 Bytes'

    const k = 1024
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']

    const i = Math.floor(Math.log(bytes) / Math.log(k))

    return parseFloat((bytes / Math.pow(k, i)).toFixed(decimals)) + ' ' + sizes[i]
  },

  formatMicroSecondes (microseconds, decimals = 2) {
    if (microseconds === 0) return '0 µs'

    const k = 1000
    const sizes = ['µs', 'ms', 's']

    const i = Math.floor(Math.log(microseconds) / Math.log(k))

    return parseFloat((microseconds / Math.pow(k, i)).toFixed(decimals)) + ' ' + sizes[i]
  },

  convertArraysOfBytes (arraysOfBytes) {
    // get the minimun value of all arrays
    const minValue = Math.min.apply(null, arraysOfBytes.flat().filter(Boolean))

    const k = 1024
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']

    const i = Math.floor(Math.log(minValue) / Math.log(k))

    // convert every value of the input arrays to the most readable unit of the minimum value
    const convertedValues = []
    arraysOfBytes.forEach(arr => {
      convertedValues.push(arr.map(bytes => parseFloat((bytes / Math.pow(k, i)).toFixed(2))))
    })

    // get the corresponding unit
    const unit = sizes[i]

    return {
      convertedValues,
      unit
    }
  },

  // functions to handle token in development mode and external users analytics service: does not impact OVH prod
  getTokenFromCookie () {
    return Cookies.get('owsToken')
  },

  removeTokenFromCookie () {
    Cookies.remove('owsToken')
  },

  setTokenInCookie (token, expiration) {
    Cookies.set('owsToken', token, {
      expires: expiration
    })
  },

  async generateToken () {
    const ovh = require('ovh')({
      endpoint: process.env.VUE_APP_ENDPOINT,
      appKey: process.env.VUE_APP_KEY,
      appSecret: process.env.VUE_APP_SECRET,
      consumerKey: process.env.VUE_APP_CONSUMER_KEY
    })

    const hosting = process.env.VUE_APP_HOSTING

    try {
      const token = await ovh.requestPromised('GET', `/hosting/web/${hosting}/userLogsToken`)
      return token
    } catch (e) {
      console.log(e)
    }
  }

}
